unified_ecommerce/test_utils.py

```diff
@@ -138,3 +223,211 @@ class PickleableMock(Mock):
+
+    Set viewset_class, factory_class and queryset to the appropriate values for your
+    viewset. `list_url` should be the root URL for the list view, and the `object_url` should
+    be contain a format string for individual objects in the viewset. DRF convention is that
+    the URLs remains the same but will get used with different HTTP verbs according
+    to the test being run (i.e. update sends a PATCH, etc.).
+
+    The tests that perform writes to the dataset require subclassing. They're more
+    abstractions of the HTTP request than an actual test; you'll need to provide some
+    logic for the test to check the operation succeeded with the model you're using.
+    """
+
+    viewset_class = None
+    factory_class = None
+    queryset = None
+
+    # URLs for these actions
+    list_url = None
+    object_url = None
+
+    def _test_retrieval(self, api_client, url, url_name, **kwargs):
+        """
+        Test that hitting the specified URL works with the specified client.
+
+        Args:
+        - api_client (APIClient): the client to use
+        - url (str): the URL to test with
+        - url_name (str): the name of the URL (used for the skip message if it's not defined)
+
+        Keyword Args:
+        - test_non_authenticated (bool): whether or not this should expect a 403
+
+        Returns:
+        - response (Response): the response from the API
+        """
+        if not url:
+            exception_string = f"{url_name} is not defined"
+            raise ViewSetNotConfiguredError(exception_string)
+
+        response = api_client.get(url)
+        assert response.status_code == 403 if kwargs["test_non_authenticated"] else 200
+        return response
+
+    def test_get_queryset(self):
+        """Test that the viewset returns the correct queryset."""
+        queryset = self.viewset_class().get_queryset()
+        assert queryset.count() == self.queryset.count()
+
+    def test_get_serializer_class(self):
+        """Test that the viewset returns the correct serializer class."""
+        serializer_class = self.viewset_class().get_serializer_class()
+        assert serializer_class == self.viewset_class.serializer_class
+
+    @pytest.mark.parametrize("isLoggedIn", [True, False])
+    def test_list(self, isLoggedIn, client, user_client):
+        """
+        Test that the viewset can list objects.
+
+        Args:
+        - isLoggedIn (bool): whether or not the client is logged in
+        - client (APIClient): the client to use for non-logged in requests
+        - user_client (APIClient): the client to use for logged in requests
+        """
+
+        response = self._test_retrieval(
+            user_client if isLoggedIn else client,
+            self.list_url,
+            "list_url",
+            test_non_authenticated=not isLoggedIn,
+        )
+
+        if isLoggedIn:
+            assert "count" in response.data
+            assert response.data["count"] == self.queryset.count()
```

Comment:
What are valid assertions if they're not logged in? Or is that not something easily generalized?

Reply:
I think it veers into the not easily generalized category - depending on the API you may see all of the dataset, a subset of it, or everything. (We definitely do the former and latter but not so much the middle case.) 

This case is potentially there too for logged in users, but I think it's more common in our case that you'll just see everything that matches the queryset provided as a logged in user.

Reply:
Makes sense, I sort of figured that was the case.